### ISSUE TITLE
Model Fit deadlocks when training in SageMaker with PipeModeDataset

### ISSUE THREAD

**Issue by murphycrosby** [philipperemy/keras-tcn#186]
Model Fit deadlocks when training on SageMaker with PipeModeDataset. The model works fine when you swap out PipeModeDataset with tf.data.TFRecordDataset(files)

Example TCN Model
`
i = Input(batch_shape=(None, 10, input_dim))
o = TCN(return_sequences=False)(i)  # The TCN layers are here.
o = Dense(1)(o)
m = Model(inputs=[i], outputs=[o])
m.compile(optimizer=args.optimizer, loss='mse')
tcn_full_summary(m, expand_residual_blocks=False)

ds = PipeModeDataset(channel='train', record_format='TFRecord')

ds = ds.map(parse, num_parallel_calls=10)
ds = ds.prefetch(10)
ds = ds.repeat(10)
ds = ds.batch(25, drop_remainder=True)

for row in ds:
    print(f'x_shape: {row[0]["data"].numpy().shape}')
    print(row[0]["data"].numpy())
    print(f'y_shape: {row[1].numpy().shape}')
    print(f'y: {row[1].numpy()}')
     break
m.fit(ds, epochs=10)
`

tensorflow==2.3.1

**Comment by philipperemy** [comment]
@murphycrosby have you ever had the chance to try on a more recent version of TF?

**Comment by krzim** [comment]
@murphycrosby Have you tried swapping out the TCN for another model to see if the issue persists? I'd be curious to see if it were actually related to the TCN training in particular or if it's an oddity related to PipeModeDataset. I would also like to see how you are setting up model training in SageMaker and what instance, options, etc. you are using.

It sounds sort of similar to this issue [here](https://github.com/aws/sagemaker-tensorflow-extensions/issues/46) which is external to the TCN portion of the code. I don't know that it matters much for a normal TFRecordsDataset but order of operations might matter on the PipeModeDataset. You could try rearranging the parse/prefetch/batch ops to match the AWS example [here](https://github.com/aws/amazon-sagemaker-examples/blob/188b0ccd1f38d387298c444e634c0e730b67b12d/sagemaker-python-sdk/keras_script_mode_pipe_mode_horovod/source_dir/cifar10_keras_main.py#L139).

Lastly, you could swap to the new [FastFile Mode](https://aws.amazon.com/about-aws/whats-new/2021/10/amazon-sagemaker-fast-file-mode/) with something like this example [here](https://towardsdatascience.com/amazon-sagemaker-fast-file-mode-d12829479c39).

**Comment by philipperemy** [comment]
I'll close this issue because it's obsolete.

AWS itself now tends to recommend FastFile or FSx instead of PipeMode for TensorFlow pipelines.
